File: 0004 Median of Two Sorted Arrays/solution.py
#!python3

class Solution:
    def findMedianSortedArrays(self, nums1, nums2):
        if len(nums1) > len(nums2):
            return self.findMedianSortedArrays(nums2, nums1)
        m, n = len(nums1), len(nums2)
        imin, imax = 0, m 
        
        while imin <= imax:
            i = (imin+imax)//2     #nums1的切分点
            j = (m+n+1) // 2 - i   #nums2的切分点
            
            if i<m and nums2[j-1] > nums1[i]: #i 太小
                imin = i + 1
            elif i>0 and nums1[i-1] > nums2[j]: #i 太大
                imax = i - 1
            else:
                if i==0: max_of_left = nums2[j-1]
                elif j==0: max_of_left = nums1[i-1]
                else:    max_of_left = max(nums1[i-1], nums2[j-1]);
                
                if (m + n) % 2 == 1:   #奇数
                    return max_of_left
                
                if i == m:   min_of_right = nums2[j]
                elif j == n: min_of_right = nums1[i]
                else:
                    min_of_right = min(nums1[i], nums2[j]);
                
                return (max_of_left + min_of_right) / 2.0

if __name__ == "__main__":
    n1, n2 = [1, 2], [3, 4]
    print(Solution().findMedianSortedArrays(n1, n2) == 2.5)

File: 0004 Median of Two Sorted Arrays/solution.cpp
#include <bits/stdc++.h>
using namespace std;

 
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        if(nums1.size()>nums2.size())
            return findMedianSortedArrays(nums2, nums1);
        
        int len=nums1.size()+nums2.size();
        int cutL=0, cutR=nums1.size()-1;
        int cut1, cut2;
        
        while(true){
            cut1 = cutL+(cutR-cutL)/2;
            cut2 = len/2-cut1;
            int L1 = (cut1==0) ? INT_MIN : nums1[cut1-1];
            int L2 = (cut2==0) ? INT_MIN : nums2[cut2-1];
            int R1 = (cut1==nums1.size()) ? INT_MAX : nums1[cut1];
            int R2 = (cut2==nums2.size()) ? INT_MAX : nums2[cut2];
            if(L1>R2) // to left
                cutR = cut1-1;
            else if(L2>R1) //to right
                cutL = cut1+1;
            else{
                if(len%2==0) // even , choose (k-1, k)/2
                    return (max(L1, L2) + min(R1, R2))/2.0;
                else  //odd
                    return min(R1, R2);
            }
        }
    }
};

int main()
{
    Solution s;
    vector<int> n1{1,3}, n2{2};
    assert( (s.findMedianSortedArrays(n1, n2) == 2.0) );
    return 0;
}